main/services/devon-instances/devon-instances.service.ts

```diff
@@ -1,99 +1,150 @@
-                  singlepath = singlepath.replace('/', '');
-                  singlepath = singlepath.replace('/', ':/');
-                  singlepath = singlepath.replace(/\//g, path.sep);
-                }  
-                const instance: IdeDistribution = {
-                  id: singlepath,
-                  ideConfig: {
-                    version: '',
-                    basepath: singlepath,
-                    commands: path.resolve(singlepath, 'scripts', 'command'),
-                    workspaces: path.resolve(singlepath, 'workspaces'),
-                  },
-                };
-                exec(
-                  'devon -v',
-                  { cwd: path.resolve(singlepath, 'scripts') },
-                  (_: unknown, stdout: string) => {
-                    instance.ideConfig.version = stdout;
-                  }
-                );
-                instances.distributions.push(instance);
-              }
-            }
-            resolve(instances);
-          }
-        }
-      );
-    });
-    return instancesDirReader;
-  }
-}
+import fs from 'fs';
+import path from 'path';
+import platform from 'os';
+import {
+  DevonfwConfig,
+  IdeDistribution,
+} from '../../models/devonfw-dists.model';
+import * as child from 'child_process';
+import { ProjectDetails } from '../../models/project-details.model';
+
+const exec = child.exec;
+
+export class DevonInstancesService {
+  private devonFilePath = path.resolve(
+    platform.homedir(),
+    '.devon',
+    'projectinfo.json'
+  );
+
+  getAvailableDevonIdeInstances(): Promise<number> {
+    let instanceCount = 0;
+    const promiseInstances = [];
+    const dirReader = new Promise<number>((resolve, reject) => {
+      this.getAllUserCreatedDevonInstances().then(
+        (instances: DevonfwConfig) => {
+          for (const distribution of instances.distributions) {
+            if (distribution.id) {
+              promiseInstances.push(this.getInstances(distribution.id));
+            }
+          }
+          if (promiseInstances.length) {
+            Promise.all(promiseInstances)
+              .then((results) => {
+                for (const result of results) {
+                  instanceCount = instanceCount + result;
+                }
+                resolve(instanceCount);
+              })
+              .catch((error) => {
+                console.log(error);
+                reject(error);
+              });
+          }
+        }
+      );
+    });
+    return dirReader;
+  }
+
+  getInstances(instancepath: string): Promise<number> {
+    const devonInstances = new Promise<number>((resolve, reject) => {
+      fs.readdir(path.resolve(instancepath, 'workspaces'), (error, files) => {
+        if (error) reject(resolve(0));
+        if (files) {
+          resolve(files.length);
+        }
+      });
+    });
+    return devonInstances;
+  }
+
+  getAllUserCreatedDevonInstances(): Promise<DevonfwConfig> {
+    let paths = [];

```

Comment:
Here the same. The function contains lots of sentences, difficulting the readability